src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs

```diff
@@ -1208,82 +1122,126 @@ public static string GetPSModulePath(PSModulePathScope scope)
+            List<string> finalModulePath = new List<string>();
+            AddPath(finalModulePath, personalModulePathToUse);
+            AddPath(finalModulePath, sharedModulePath);
+            AddPath(finalModulePath, systemModulePathToUse);
+
+            // If the existing module path if provided, we add it to the end of the list.
+            if (currentProcessModulePath != null)
             {
-                // Handle the default case...
-                if (string.IsNullOrEmpty(hkcuUserModulePath)) // EVT.User does Not exist -> set to <SpecialFolder.MyDocuments> location
-                {
-                    currentProcessModulePath = personalModulePath; // = SpecialFolder.MyDocuments + Utils.ProductNameForDirectory + Utils.ModuleDirectory
-                }
-                else // EVT.User exists -> set to EVT.User
-                {
-                    currentProcessModulePath = hkcuUserModulePath; // = EVT.User
-                }
+                AddPath(finalModulePath, currentProcessModulePath);
+            }
+
+            return string.Join(Path.PathSeparator, finalModulePath);
+        }
 
-                if (string.IsNullOrEmpty(currentProcessModulePath))
+        /// <summary>
+        /// Adds the path if not present to the final module path list. Each
+        /// path is split by the path separator and added to the final list.
+        /// </summary>
+        /// <param name="finalModulePath">The final module path list.</param>
+        /// <param name="path">The path to add.</param>
+        private static void AddPath(List<string> finalModulePath, ReadOnlySpan<char> path)
+        {
+            ReadOnlySpan<char> dirSeparatorChars = stackalloc char[] {
+                Path.DirectorySeparatorChar,
+                Path.AltDirectorySeparatorChar
+            };
+
+#if UNIX
+            IEqualityComparer<char>? valueComparer = null;
+#else
+            IEqualityComparer<char> valueComparer = new WindowsPathEnvComparer();
+#endif
+
+            foreach (Range segment in path.Split(Path.PathSeparator))
+            {
+                // We trim any surrounding whitespace and trailing directory separators
+                // to canonicalize the path. The comparer on Windows will also treat
+                // '/' and '\' as the same and in a case insensitive operation.
+                scoped ReadOnlySpan<char> entry = path[segment].Trim();
+                if (entry.Length == 0)
                 {
-                    currentProcessModulePath ??= string.Empty;
+                    continue;
                 }
+#if UNIX
+                // Don't trim if it's just '/' as that's a valid entry on *nix.
+                else if (!(entry.Length == 1 && entry[0] == Path.DirectorySeparatorChar))
+#else
                 else
+#endif
                 {
-                    currentProcessModulePath += Path.PathSeparator;
+                    entry = entry.TrimEnd(dirSeparatorChars);
```

Comment:
Fair point I’ll use the params overload which I think is the same one used but avoids the span allocation.

Reply:
Please see my comment about [TrimEndingDirectorySeparator()](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IO/Path.cs,957)

Reply:
Ah there reason why I did this is because the `params` overload is for a string, we are operating on a `ReadOnlySpan` and [MemoryExtensions.TrimEnd](https://learn.microsoft.com/en-us/dotnet/api/system.memoryextensions.trimend?view=net-8.0) doesn't have such an overload.

Reply:
> I suggest to use more simple code with TrimEndingDirectorySeparator() here. It is well tested (I see one issue in your code for Windows)

What's the issue for Windows?

I'll wait until the WG weighs in to see whether they prefer the current code that trims all trailing dir path separators or whether to swap to `TrimEndingDirectorySeparator` which only trims the last.

Reply:
```c#
                // On Windows we always want to trim '/' or '\'. Only exception
                // is to keep one trailing char if the path is the drive root
                // 'C:\+'.
                entry = entry.Length > 1 && entry[1] == Path.VolumeSeparatorChar && trimmedEntry.Length == 2
                    ? entry.Slice(0, 3)
                    : trimmedEntry;
```
The code is not correct.

------------

You will wait indefinitely.
I suggest making the code as simple as possible and accepting it now.
Since we are talking about an extreme case, you can create an issue for it and make changes to the code after additional discussion if it really turns out to be useful for many users.

Reply:
> You will wait indefinitely.

I'm assuming I'm waiting indefinitely to get this reviewed by the WG anyway :) I appreciate you reviewing these PRs though and I'm not trying to be obstructionist it's just if they have to review this anyway I prefer to wait for the final verdict before changing things.

I have picked up the bug if it's `C:` then the `Slice` will go out of bounds and will fix up. Not sure if that's the incorrect part you are referring to though.